### PR TITLE
Remove GTNode

### PR DESCRIPTION
**Pull request by atgeirr**
This include #946, but also does some other stuff that needs reviewing:
 - Adding a getGroups2() method to Schedule, similar to the getWells2() method. It returns a vector of pointers instead of values. I think getWells2() should be changed to return pointers as well to avoid performance pessimization, but that is not done in this PR.
 - maxGroupSize() in ArrayDimChecker() is redone using getGroups2(). In my understanding it is not necessary to use a recursive or hierarchical algorithm for this.

### CODE REVIEW

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
> Adding a getGroups2() method to Schedule, similar to the getWells2() method. It returns a vector of pointers instead of values.

Personally I would prefer to go the other way - i.e. remove the `getWells2()` method. The `Schedule` api is large and unwieldy already.

**Comment by atgeirr** [comment]
> Personally I would prefer to go the other way - i.e. remove the `getWells2()` method. The `Schedule` api is large and unwieldy already.

Let's discuss that later, and merge the other PR instead...

File: src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 1]
```diff
@@ -2083,6 +2056,19 @@ void Schedule::handleGRUPTREE( const DeckKeyword& keyword, size_t currentStep, c
...
     }
 
+    std::vector<Group2*> Schedule::getGroups2(size_t timeStep) const {
+        std::vector<Group2*> groupsAtTimestep;
+        if (timeStep >= this->m_timeMap.size())
+            throw std::invalid_argument("timeStep argument beyond the length of the simulation");
+
+        for (const auto& dynamic_pair : this->groups) {
+            auto& group_ptr = dynamic_pair.second.get(timeStep);
+            if (group_ptr)
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: Look ma - no braces!

Comment (atgeirr) [reply src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: Yes I don't like it, but I try to preserve existing style. The function is identical to getWells2 apart from replacing `this->wells_static` with `this->groups`.

**Comment by bska** [comment]
> `maxGroupSize()` in `ArrayDimChecker` is redone using `getGroups2`. In my understanding it is not necessary to use a recursive or hierarchical algorithm for this.

This captures the essence of the current code, but the current code is wrong (it well groups only).  The `maxGroupSize()` function is supposed to return the maximum group size at a particular time step across all groups defined at that time step.  For well groups, the group size is the number of wells in the group but for node groups the group size is the number of child groups.

I didn't notice this change in semantics when the current version was added in commit 7c856f68 (PR #900).

In hindsight, and in terms of `Group2` you could probably implement `maxGroupSize()` as
```C++
int maxGroupSize(const Schedule& sched, const std::size_t step)
{
    int nwgmax = 0;

    for (const auto& gnm : sched.groupNames(step)) {
        const auto& grp = sched.getGroup2(gnm, step);
        const auto  gsz = grp.wellgroup()
            ? grp.numWells() : grp.groups().size();

        nwgmax = std::max(nwgmax, static_cast<int>(gsz));
    }

    return nwgmax;
}
```

**Comment by joakim-hove** [comment]
> This captures the essence of the current code, but the current code is wrong 

*Current code* in this context is that master, or the PR? In case of master I guess you are implying that there is a bug in master which should be fixed - irrespective of the PR?

**Comment by bska** [comment]
> > This captures the essence of the current code, but the current code is wrong
> 
> _Current code_ in this context is that master, or the PR?

It's the current master sources.  Sorry for being imprecise.

> In case of master I guess you are implying that there is a bug in master which should be fixed - irrespective of the PR?

Yes, upon reviewing the master sources now I do believe there is a problem in the `maxGroupSize()` function.  I'm sorry this wasn't flagged by the unit tests.

**Comment by joakim-hove** [comment]
> It's the current master sources. Sorry for being imprecise.

OK - "good" - then we must fix it. Do you want to submit a PR - or should I do it?

**Comment by bska** [comment]
> > It's the current master sources. Sorry for being imprecise.
> 
> OK - "good" - then we must fix it. Do you want to submit a PR - or should I do it?

I'm on it.  Update later today.

**Comment by bska** [comment]
> I'm on it (fixing incorrect "effective nwgmax" calculation). Update later today.

Please see PR #991.

**Comment by blattms** [comment]
Shall we close this? Seems rather old and stale.

**Comment by bska** [comment]
> Shall we close this?

I think so, but @atgeirr should comment first.  It's not clear to me what role the `GTNode` plays anymore.

**Comment by atgeirr** [comment]
> I think so, but @atgeirr should comment first. It's not clear to me what role the `GTNode` plays anymore.

The huge performance problem that was the start of this was addressed back in 2019, this was just a followup, realizing that we could get rid of the whole class. I have not looked at it since then, but the class still exists and should probably still be removed. However, any concrete changes to other code will need to be redone. It will not be difficult, so there is no point in rebasing this. Closing.

**Pull request closed without merging**